vyper/parser/parser.py

```diff
@@ -86,20 +87,36 @@ def parse_external_interfaces(external_interfaces, global_ctx):
     return external_interfaces
 
 
-def parse_other_functions(
-    o, otherfuncs, sigs, external_interfaces, global_ctx, default_function, is_contract_payable,
-):
+def parse_other_functions(o, otherfuncs, sigs, external_interfaces, global_ctx, default_function):
+    # check for payable/nonpayable external functions to optimize nonpayable assertions
+    func_types = [i._metadata["type"] for i in global_ctx._defs]
+    mutabilities = [i.mutability for i in func_types if i.visibility == FunctionVisibility.EXTERNAL]
+    has_payable = next((True for i in mutabilities if i == StateMutability.PAYABLE), False)
+    has_nonpayable = next((True for i in mutabilities if i != StateMutability.PAYABLE), False)
```

Comment:
Seems clearer that this be a straight `bool`
```suggestion
    has_payable = next((True for i in mutabilities if i == StateMutability.PAYABLE), None) is None
    has_nonpayable = next((True for i in mutabilities if i != StateMutability.PAYABLE), None) is None
```